File: py/scoring.py
import argparse
import bisect
from collections import deque as Queue
import tqdm


def _parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('input')
    parser.add_argument('output')
    return parser.parse_args()


args = _parse_args()


class Street(object):
    def __init__(self, name, length, begin, end, idx):
        self.name = name
        self.length = length
        self.begin = begin
        self.end = end
        self.idx = idx
        self.queue = Queue()
    
    def push(self, v_idx):
        self.queue.append(v_idx)

    def pop(self):
        return self.queue.popleft()

    def empty(self):
        return not bool(self.queue)


class Intersection(object):
    def __init__(self):
        self.input = set()
        self.output = set()
        self.schedule = None
        self.period = None
        self.schedule_prefix = None

    def add_in_street(self, idx):
        assert idx not in self.input
        self.input.add(idx)
    
    def add_out_street(self, idx):
        assert idx not in self.output
        self.output.add(idx)

    def set_schedule(self, schedule):
        self.schedule = schedule
        self.period = sum([t for _, t in self.schedule])
        self.schedule_prefix_sum = [0]
        for _, t in self.schedule:
            self.schedule_prefix_sum.append(self.schedule_prefix_sum[-1] + t)


    def get_current_street(self, timestamp):
        if self.schedule is None:
            return None

        if self.period == 0:
            return None
        reminder = timestamp % self.period
        schedule_idx = bisect.bisect(self.schedule_prefix_sum, reminder) - 1
        
        return self.schedule[schedule_idx][0]



class Vehicle(object):
    def __init__(self, path):
        self.path = path
        self.street_rel_idx = 0

    def move(self):
        self.street_rel_idx += 1
    
    def get_current_street(self):
        assert 0 <= self.street_rel_idx < len(self.path)
        return self.path[self.street_rel_idx]

    def is_finished(self):
        return self.street_rel_idx == len(self.path) - 1


streets = []
intersections = []
vehicles = []

STREET_NAME_TO_IDX = {}

with open(args.input) as fp:
    finput = lambda: fp.readline().strip()
    total_duration, num_intersections, num_streets, num_vehicles, finish_bonus = map(int, finput().split())
    intersections = [Intersection() for _ in range(num_intersections)]

    for idx in range(num_streets):
        begin, end, name, length = finput().split()
        begin = int(begin)
        end = int(end)
        length = int(end)
        streets.append(Street(name=name, length=length, begin=begin, end=end, idx=idx))
        intersections[begin].add_out_street(idx)
        intersections[end].add_in_street(idx)
        STREET_NAME_TO_IDX[name] = idx


    for v_idx in range(num_vehicles):
        p_names = finput().split()
        p = int(p_names[0])
        names = p_names[1:]
        path = []
        assert len(names) == p
        for name in names:
            s_idx = STREET_NAME_TO_IDX[name]
            path.append(s_idx)
        vehicles.append(Vehicle(path))
        streets[path[0]].push(v_idx)


with open(args.output) as fp:
    finput = lambda: fp.readline().strip()
    a = int(finput())
    assert 0 <= a <= num_intersections
    for _ in range(a):
        i_idx = int(finput())
        assert 0 <= i_idx < len(intersections)
        intersection = intersections[i_idx]
        assert intersection.schedule is None
        num_incoming_streets = int(finput())
        assert 0 <= num_incoming_streets <= len(intersection.input)
        schedule = []
        for street_rel_idx in range(num_incoming_streets):
            name, t = finput().split()
            t = int(t)
            assert 1 <= t <= total_duration
            assert name in STREET_NAME_TO_IDX
            street_idx = STREET_NAME_TO_IDX[name]
            assert street_idx in intersection.input
            schedule.append((street_idx, t))
        intersection.set_schedule(schedule)


CAR_EVENTS = [[] for _ in range(total_duration + 1)]

total_score = 0

for timestamp in tqdm.trange(0, total_duration + 1):
    print("TIMESTAMP", timestamp)
    # lookup all cat events and push them
    for event in CAR_EVENTS[timestamp]:
        car_idx, street_idx = event
        if vehicles[car_idx].is_finished():
            print('TIMESTAMP', timestamp, 'CAR', car_idx, 'IS FINISHED', 'SCORE', finish_bonus + (total_duration - timestamp))
            total_score += finish_bonus + (total_duration - timestamp)
        else:
            streets[street_idx].push(car_idx)
            print(vehicles[car_idx].path)
            print('TIMESTAMP', timestamp, 'car', car_idx, 'pushed', street_idx, streets[street_idx].name)

    for intersection_idx, intersection in enumerate(intersections):
        street_idx = intersection.get_current_street(timestamp)
        if street_idx is None:
            continue
        # print(intersection_idx, street_idx, streets[street_idx].name)
        if not streets[street_idx].empty():
            car_idx = streets[street_idx].pop()
            vehicles[car_idx].move()
            # assert not vehicles[car_idx].is_finished()
            next_street_idx = vehicles[car_idx].get_current_street()
            length = streets[next_street_idx].length
            if timestamp + length <= total_duration:
                CAR_EVENTS[timestamp + length].append((car_idx, next_street_idx))
                print('TIMESTAMP', timestamp, 'car', car_idx, 'next street', next_street_idx, streets[next_street_idx].name, length)
            else:
                print('Too long', timestamp + length, car_idx)

print(total_score)

File: cpp/src/scoring.cpp
#include "game.hpp"
#include "scoring.hpp"
#include "solution.hpp"
#include <cstdint>

#include <queue>
#include <algorithm>
#include <iostream>


struct StreetImpl {
    const Street& data_;

    explicit StreetImpl(const Street& data)
        : data_(data) 
    {
    }

    void push(int v_idx) {
        queue.push(v_idx);
    }

    int pop() {
        if (empty()) {
            throw "EMPTY";
        }
        auto res = queue.front();
        queue.pop();
        return res;
    }

    bool empty() {
        return queue.empty();
    }

    std::queue<int> queue;
};


struct IntersectionImpl {

    const Intersection& data_;
    std::vector<std::pair<int, int>> schedule;
    std::vector<int64_t> prefix_sum;
    int64_t period = 0;

    explicit IntersectionImpl(const Intersection& data)
        : data_(data) 
    {
    }


    void set_schedule(const std::vector<std::pair<int, int>>& schedule_in) {
        schedule = schedule_in;
        prefix_sum.push_back(0);
        for (const auto& element : schedule) {
            prefix_sum.push_back(prefix_sum.back() + element.second);
        }
        period = prefix_sum.back();
    }

    int get_current_street(int64_t timestamp) {
        if (schedule.empty()) {
            return -1;
        }
        if (period == 0) {
            return -1;
        }
        int64_t reminder = timestamp % period;
        auto schedule_idx = std::upper_bound(prefix_sum.begin(), prefix_sum.end(), reminder) - prefix_sum.begin() - 1;
        return schedule[schedule_idx].first;
    }
};

struct CarImpl {
    const Car& data_;
    int street_rel_idx = 0;
    int64_t arrive_timestamp = 0;

    explicit CarImpl(const Car& data)
        : data_(data)
    {
    }

    void move() {
        ++street_rel_idx;
    }

    int get_current_street() {
        if (!(0 <= street_rel_idx && street_rel_idx < data_.path_streets.size())) {
            throw "Error";
        }
        return data_.path_streets[street_rel_idx];
    }

    bool is_finished() {
        return street_rel_idx + 1 == data_.path_streets.size();
    }
};



void DebugInfo::Print() {
    std::cerr << "Cars not finished: " << cars_not_finished << std::endl;
    for (const auto& event : events) {
        std::cerr << 
            "car\t" << event.car_idx << 
            "\tstreet\t" << event.street_idx << 
            "\tintersection\t" << event.intersection_idx << 
            "\tarrive_ts\t" << event.arrive_timestamp << 
            "\twait_time\t" << event.wait_time << 
            "\tis_first\t" << int(event.is_first) << 
            std::endl;
    }
}

int64_t Score(const Game& input, const GameSolution& solution, DebugInfo* debug_info_sink) {
    DebugInfo debug_info;

    std::vector<std::vector<std::pair<int, int>>> car_events(input.D + 1);
    int64_t total_score = 0;
    std::vector<IntersectionImpl> intersections;
    std::vector<CarImpl> cars;
    std::vector<StreetImpl> streets;

    for (const auto& data : input.intersections) {
        intersections.emplace_back(data);
    }

    for (const auto& data : input.streets) {
        streets.emplace_back(data);
    }

    int idx = 0;
    for (const auto& data : input.cars) {
        cars.emplace_back(data);
        auto& car = cars.back();
        streets[car.data_.path_streets[0]].push(idx);
        car.arrive_timestamp = 0;
        ++debug_info.cars_not_finished;
        ++idx;
    }

    idx = 0;
    for (const auto& schedule : solution.intersection_assignment) {
        intersections[idx].set_schedule(schedule);
        ++idx;
    }

    for (int64_t timestamp = 0; timestamp <= input.D; ++timestamp) {
        for (const auto& event : car_events[timestamp]) {
            auto car_idx = event.first;
            auto street_idx = event.second;
            if (cars[car_idx].is_finished()) {
                total_score += input.F + (input.D - timestamp);
                --debug_info.cars_not_finished;
            } else {
                streets[street_idx].push(car_idx);
                cars[car_idx].arrive_timestamp = timestamp;
            }
        }

        for (int intersection_idx = 0; intersection_idx < intersections.size(); ++intersection_idx) {
            auto& intersection = intersections[intersection_idx];
            auto street_idx = intersection.get_current_street(timestamp);
            if (street_idx == -1) {
                continue;
            }
            if (!streets[street_idx].empty()) {
                auto car_idx = streets[street_idx].pop();
                DebugInfo::Event event;
                
                event.car_idx = car_idx;
                event.street_idx = street_idx;
                event.intersection_idx = intersection_idx;
                event.arrive_timestamp = cars[car_idx].arrive_timestamp;
                event.wait_time = timestamp - event.arrive_timestamp;
                event.is_first = (cars[car_idx].street_rel_idx == 0);
                debug_info.events.push_back(std::move(event));

                cars[car_idx].move();
                auto next_street_idx = cars[car_idx].get_current_street();
                auto length = streets[next_street_idx].data_.L;
                if (timestamp + length <= input.D) {
                    car_events[timestamp + length].push_back({car_idx, next_street_idx});
                }
            }
        }
    }

    if (debug_info_sink != nullptr) {
        *debug_info_sink = std::move(debug_info);
    }

    return total_score;
}